File: GrokkingDP/5/ShortestCommonSuperseq.cpp
#include <iostream>
#include <vector>

enum class Method { BRUTE_LCS, BRUTE, MEM, BU };

int ShortestCommonSuperseq(std::string seq1, std::string seq2, Method m);
int BruteLCS(std::string seq1, std::string seq2, size_t i1, size_t i2);
int Brute(std::string seq1, std::string seq2, size_t i1, size_t i2);
int Memoize(std::string seq1, std::string seq2, size_t i1, size_t i2,
            std::vector<std::vector<int>>& dp);

int main(void) {
  Method m = Method::BRUTE;
  std::cout << ShortestCommonSuperseq("abcf", "bdcf", m) << ":"
            << ShortestCommonSuperseq("dynamic", "programming", m);
  return 0;
}

int ShortestCommonSuperseq(std::string seq1, std::string seq2, Method m) {
  int longest_subseq;
  switch (m) {
    case Method::BRUTE_LCS:
      longest_subseq = BruteLCS(seq1, seq2, 0, 0);
      return (seq1.length() - longest_subseq) +
             (seq2.length() - longest_subseq) + longest_subseq;
    case Method::BRUTE:
      return Brute(seq1, seq2, 0, 0);
    case Method::MEM:
    case Method::BU:
      return 0;
  }
}

int BruteLCS(std::string seq1, std::string seq2, size_t i1, size_t i2) {
  if (i1 == seq1.length() || i2 == seq2.length()) return 0;
  if (seq1[i1] == seq2[i2]) return 1 + BruteLCS(seq1, seq2, i1 + 1, i2 + 1);
  return std::max(BruteLCS(seq1, seq2, i1 + 1, i2),
                  BruteLCS(seq1, seq2, i1, i2 + 1));
}

int Brute(std::string seq1, std::string seq2, size_t i1, size_t i2) {
  if (i1 == seq1.length() || i2 == seq2.length())
    return std::max(seq1.length() - i1, seq2.length() - i2);
  if (seq1[i1] == seq2[i2]) return 1 + Brute(seq1, seq2, i1 + 1, i2 + 1);
  return 1 +
         std::min(Brute(seq1, seq2, i1 + 1, i2), Brute(seq1, seq2, i1, i2 + 1));
};

int Memoize(std::string seq1, std::string seq2, size_t i1, size_t i2,
            std::vector<std::vector<int>>& dp) {
}

File: GrokkingDP/5/shortest_common_superseq.py
#  Problem Statement #
#  Given two sequences ‘s1’ and ‘s2’, write a method to find the length of the shortest sequence which has ‘s1’ and ‘s2’ as subsequences.

#  Example 1:

#  Input: s1: "abcf" s2:"bdcf"
#  Output: 5
#  Explanation: The shortest common super-sequence (SCS) is "abdcf".
#  Example 2:

#  Input: s1: "dynamic" s2:"programming"
#  Output: 15
#  Explanation: The SCS is "dynprogrammicng".

def bu(s1: str, s2: str) -> int:
    """
    >>> bu('abcf', 'bdcf')
    5
    >>> bu('dynamic', 'programming')
    15
    """
    A, B = len(s1), len(s2)
    if B > A:
        A, B = B, A
        s1, s2 = s2, s1
    dp = [[0 for _ in range(B)] for _ in range(2)]
    for i in range(A):
        for j in range(B):
            if s1[i] == s2[j]: dp[1][j] = dp[0][j-1] + 1
            else: dp[1][j] = max(dp[0][j], dp[1][j-1])
        dp[0] = dp[1][:]
    return A + B - dp[1][-1]

if __name__ == '__main__':
    import doctest
    doctest.testmod()
